graph/api/src/main/scala/org/apache/spark/graph/api/CypherSession.scala

```diff
@@ -0,0 +1,138 @@
+   * [[RelationshipFrame]], respectively.
+   *
+   * @param nodes         NodeFrames that define the nodes in the graph
+   * @param relationships RelationshipFrames that define the relationships in the graph
+   * @since 3.0.0
+   */
+  def createGraph(
+      nodes: java.util.List[NodeFrame],
+      relationships: java.util.List[RelationshipFrame]): PropertyGraph = {
+    createGraph(JavaConverters.asScalaBuffer(nodes), JavaConverters.asScalaBuffer(relationships))
+  }
+
+  /**
+   * Creates a [[PropertyGraph]] from nodes and relationships.
+   *
+   * The given DataFrames need to adhere to the following column naming conventions:
+   *
+   * {{{
+   *     Id column:        `$ID`            (nodes and relationships)
+   *     SourceId column:  `$SOURCE_ID`     (relationships)
+   *     TargetId column:  `$TARGET_ID`     (relationships)
+   *
+   *     Label columns:    `:{LABEL_NAME}`  (nodes)
+   *     RelType columns:  `:{REL_TYPE}`    (relationships)
+   *
+   *     Property columns: `{Property_Key}` (nodes and relationships)
+   * }}}
+   *
+   * @see [[CypherSession]]
+   * @param nodes         node DataFrame
+   * @param relationships relationship DataFrame
+   * @since 3.0.0
+   */
+  def createGraph(nodes: DataFrame, relationships: DataFrame): PropertyGraph = {
+    val idColumn = CypherSession.ID_COLUMN
+    val sourceIdColumn = CypherSession.SOURCE_ID_COLUMN
+    val targetIdColumn = CypherSession.TARGET_ID_COLUMN
+
+    val labelColumns = nodes.columns.filter(_.startsWith(CypherSession.LABEL_COLUMN_PREFIX)).toSet
+    val nodeProperties = (nodes.columns.toSet - idColumn -- labelColumns)
+      .map(col => col -> col)
+      .toMap
+
+    val trueLit = functions.lit(true)
+    val falseLit = functions.lit(false)
+
+    val labelSets = labelColumns.subsets().toSet + Set.empty
+    val nodeFrames = labelSets.map { labelSet =>
+      val predicate = labelColumns
+        .map {
+          case labelColumn if labelSet.contains(labelColumn) => nodes.col(labelColumn) === trueLit
+          case labelColumn => nodes.col(labelColumn) === falseLit
+        }
+        .reduce(_ && _)
+
+      NodeFrame(nodes.filter(predicate), idColumn, labelSet.map(_.substring(1)), nodeProperties)
+    }
+
+    val relColumns = relationships.columns.toSet
+    val relTypeColumns = relColumns.filter(_.startsWith(CypherSession.LABEL_COLUMN_PREFIX))
+    val propertyColumns = relColumns - idColumn - sourceIdColumn - targetIdColumn -- relTypeColumns
+    val relProperties = propertyColumns.map(col => col -> col).toMap
+    val relFrames = relTypeColumns.map { relTypeColumn =>
+      val predicate = relationships.col(relTypeColumn) === trueLit
+
+      RelationshipFrame(
+        relationships.filter(predicate),
+        idColumn,
+        sourceIdColumn,
+        targetIdColumn,
+        relTypeColumn.substring(1),
+        relProperties)
+    }
+
+    createGraph(nodeFrames.toSeq, relFrames.toSeq)
```

Comment:
Should we cache `nodes` and `relationships` dataframe first ?
The code here filter the two dataframes multiple times.

Reply:
I think if we add some documentation about the combinatorial problem, the user can decide to cache `nodes` and `relationships` before entering that method. Otherwise, we would also cache for the case, where we just have a small set of label columns.

Reply:
I think cache won't hurt perf too much in any case. But no caching is risky. The input dataframe computation may consume massive time.
@meng what do you think ?

Reply:
You probably want to know what @mengxr thinks ;)